client/hwlib/src/collectors/hardware_info.rs

```diff
@@ -0,0 +1,324 @@
+    let board_info = &smbios_data.collect::<smbioslib::SMBiosBaseboardInformation>()[0];
+
+    let manufacturer = board_info.manufacturer().to_string();
+    let product_name = board_info.product().to_string();
+    let version = board_info.version().to_string();
+
+    let board = devices::Board {
+        manufacturer,
+        product_name,
+        version,
+    };
+
+    Ok(board)
+}
+
+pub fn collect_motherboard_info_from_device_tree(
+    device_tree_filepath: &'static str,
+) -> Result<devices::Board> {
+    let base_path = std::path::Path::new(device_tree_filepath);
+
+    let manufacturer = std::fs::read_to_string(base_path.join("model"))
+        .map(|s| s.trim().to_string())
+        .unwrap_or_else(|_| "Unknown".to_string());
+
+    let product_name = std::fs::read_to_string(base_path.join("compatible"))
+        .map(|s| s.trim().to_string())
+        .unwrap_or_else(|_| "Unknown".to_string());
+
+    let version = std::fs::read_to_string(base_path.join("model"))
+        .map(|s| s.trim().to_string())
+        .unwrap_or_else(|_| "Unknown".to_string());
+
+    let board = devices::Board {
+        manufacturer,
+        product_name,
+        version,
+    };
+
+    Ok(board)
+}
+
+pub fn get_system_info(smbios_data: &smbioslib::SMBiosData) -> Result<(String, String)> {
+    let system_info = &smbios_data.collect::<smbioslib::SMBiosSystemInformation>()[0];
+    Ok((
+        system_info.product_name().to_string(),
+        system_info.manufacturer().to_string(),
+    ))
+}
+
+#[cfg(target_os = "linux")]
+/// Overwrite smbioslib::table_load_from_device to use parameters for files
+/// so we can test our code without reading smbios data from the machine that
+/// runs the test
+pub(crate) fn table_load_from_device(
+    entry_file: &'static str,
+    table_file: &'static str,
+) -> Result<smbioslib::SMBiosData, std::io::Error> {
+    let version: smbioslib::SMBiosVersion;
+    let entry_path = std::path::Path::new(entry_file);
+
+    match smbioslib::SMBiosEntryPoint64::try_load_from_file(entry_path) {
+        Ok(entry_point) => {
+            version = smbioslib::SMBiosVersion {
+                major: entry_point.major_version(),
+                minor: entry_point.minor_version(),
+                revision: entry_point.docrev(),
+            }
+        }
+        Err(err) => match err.kind() {
+            std::io::ErrorKind::InvalidData => {
+                match smbioslib::SMBiosEntryPoint32::try_load_from_file(entry_path) {
+                    Ok(entry_point) => {
+                        version = smbioslib::SMBiosVersion {
+                            major: entry_point.major_version(),
+                            minor: entry_point.minor_version(),
+                            revision: 0,
+                        }
+                    }
+                    Err(err) => return Err(err),
+                }
+            }
+            _ => return Err(err),
+        },
```

Comment:
I think it would be nice to reduce the indentation on this a little, for example making use of the [`.or_else`](https://doc.rust-lang.org/std/result/enum.Result.html#method.or_else) method available to results. Also, it's possible to return `version` from this to avoid the no-value declaration:
```rust
let version = smbioslib::SMBiosEntryPoint64::try_load_from_file(entry_path)
    .map(|entry_point| smbioslib::SMBiosVersion {
        major: entry_point.major_version(),
        minor: entry_point.minor_version(),
        revision: 0,
    })
    .or_else(|err| {
        if err.kind() != ErrorKind::InvalidData {
            return Err(err)
        }
        smbioslib::SMBiosEntryPoint32::try_load_from_file(entry_path)
            .map(|entry_point| smbioslib::SMBiosVersion { ... })
    })?;

Reply:
Updated in a7f369a